### PR TITLE
sets: add sets intersection

### PR DESCRIPTION
**Pull request by marco-m-pix4d**
Part of PCI-3790

### CODE REVIEW

File: sets/sets.go [hunk 1]
```diff
@@ -72,9 +72,29 @@ func (s *Set[T]) Remove(item T) bool {
...
+				result.items[item] = struct{}{}
+			}
+		}
+	} else {
+		for item := range x.items {
+			if s.Contains(item) {
+				result.items[item] = struct{}{}
+			}
 		}
 	}
```
Comment (odormond) [inline sets/sets.go:99]: Wouldn't it be better to just swap the variables and code the loop only once?
```suggestion
	if s.Size() > x.Size() {
		// Swap them to loop over the smallest set
		s, x = x, s
	}
	for item := range s.items {
		if x.Contains(item) {
			result.items[item] = struct{}{}
		}
	}
```

**Review by odormond — APPROVED** [review]

**Review by aliculPix4D — APPROVED** [review]
I remeber doing these exercises from Exercism :)

**Review by SpectreVert — APPROVED** [review]

**Review by iAmoric — APPROVED** [review]
Just for my understanding, what is the relation between this and [PCI-3790](https://pix4dbug.atlassian.net/browse/PCI-3790) ?

File: sets/sets.go (hunk 1 above)
Comment (marco-m-pix4d) [reply sets/sets.go:99]: This is swapping pointers (`s` is a pointer):
```
func (s *Set[T]) Intersection(...)
```

 but on the other hand, in Go, "assignment to the method receiver propagates only to callees but not to callers", so at the exit of the function, the contents of the set do remain the same. Said in another way, the code proposed is safe, although confusing.

Comment (marco-m-pix4d) [reply sets/sets.go]: Just wanted to notice that GitHub itself decided that "marco-m-pix4d marked this conversation resolved".
I did nothing 🙄.

I added a second commit, that I thinks makes the code less confusing.

**Pull request merged**